### PR TITLE
[EOC-498] Display info about items count on list view

### PR DESCRIPTION
**Pull request by adamklepacz**
Chips component was inspired by Material Design Chips 
https://material.io/components/chips/#choice-chips

Items count changes dynamically when searching.

End_of_Coffee_-_track_your_office_inventory

### CODE REVIEW

File: client/src/common/components/Chips.jsx [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import './Chips.scss';
+
+const Chips = ({ count, icon, title }) => (
+  <figure className="chips" title={title}>
```
Comment (marekrozmus) [inline client/src/common/components/Chips.jsx]: use just `div`

File: client/src/common/components/Chips.jsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import './Chips.scss';
+
+const Chips = ({ count, icon, title }) => (
```
Comment (marekrozmus) [inline client/src/common/components/Chips.jsx]: Why plural name?

File: client/src/modules/list/components/ItemsContainer.jsx [hunk 3]
```diff
@@ -85,6 +87,11 @@ class ItemsContainer extends Component {
                 id: 'list.list-item.input-find-items'
               })}
             />
+            <Chips
+              count={itemsCount}
+              icon={<ListItemsIcon />}
+              title={formatMessage({ id: 'list.items-counter-label' })}
```
Comment (Ciunkos) [inline client/src/modules/list/components/ItemsContainer.jsx]: Please parameterize the title, for example: "Showing {current} of {total} items".

File: client/src/common/components/Chips.scss [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  display: inline-flex;
+  justify-content: center;
+  align-items: center;
+  background: $gray2;
+  width: auto;
+  padding: 0 four-by(3);
+  height: 32px;
+  border-radius: 16px;
+
+  &__icon {
```
Comment (Ciunkos) [inline client/src/common/components/Chips.scss]: Personally, I would go for now icon display.

Comment (adamklepacz) [reply client/src/common/components/Chips.scss]: I left icon as an option for future use of this component

**Comment by adamklepacz** [comment]
Zrzut_ekranu_2019-09-30_o_13_55_44

File: client/src/locales/en.json [hunk 5]
```diff
@@ -167,6 +167,7 @@
   "list.items-container.filter-box": "Filter by:",
   "list.items-container.sort-box": "Sort by:",
   "list.items-container.unhandled": "Unhandled",
+  "list.items-counter-label": "Showing {displayedItemsCount} of {totalItemsCount} items",
```
Comment (Ciunkos) [inline client/src/locales/en.json]: Please investigate and use plural rules (test  case 1/1).

File: client/src/modules/list/components/ItemsContainer.jsx [hunk 6]
```diff
@@ -85,15 +90,23 @@ class ItemsContainer extends Component {
                 id: 'list.list-item.input-find-items'
               })}
             />
+            <Badge
+              count={displayedItemsCount}
+              totalCount={totalItemsCount}
+              title={formatMessage(
```
Comment (Ciunkos) [inline client/src/modules/list/components/ItemsContainer.jsx]: Please move the title composition to the badge component itself.

File: client/src/modules/list/components/ItemsContainer.jsx [hunk 7]
```diff
@@ -19,7 +20,8 @@ class ItemsContainer extends Component {
     const { items } = this.props;
 
     this.state = {
-      items
+      items,
+      displayedItemsCount: 0
```
Comment (Ciunkos) [inline client/src/modules/list/components/ItemsContainer.jsx]: displayedItem~~s~~Count?

File: client/src/modules/list/components/Items/ItemsList.jsx [hunk 8]
```diff
@@ -47,32 +80,11 @@ class ItemsList extends PureComponent {
...
-      limit.archivedLimit = currentLimit;
-    } else if (doneItems) {
-      limit.doneLimit = currentLimit;
-    } else {
-      limit.unhandledLimit = currentLimit;
-    }
-
-    updateLimit(limit, listId);
-  };
+  // TODO: Remove updateItem action
```
Comment (alemikolo) [inline client/src/modules/list/components/Items/ItemsList.jsx]: restore handleLimitUpdate is necessary to decide if an item is to be created or not (look at `shouldAnimate` helper.

File: client/src/modules/list/components/Items/ItemsList.jsx [hunk 9]
```diff
@@ -5,33 +5,61 @@ import { FormattedMessage } from 'react-intl';
...
     super(props);
 
     this.state = {
+      displayedItems: null,
       limit: DISPLAY_LIMIT
     };
   }
 
+  componentDidMount() {
+    this.handleItems();
```
Comment (alemikolo) [inline client/src/modules/list/components/Items/ItemsList.jsx]: the name does not say what the method is for

File: client/src/modules/list/components/Items/ItemsList.jsx [hunk 10]
```diff
@@ -5,33 +5,61 @@ import { FormattedMessage } from 'react-intl';
...
+  }
+
+  componentDidUpdate(previousProps) {
+    const { items: previousItems } = previousProps;
+    const { items } = this.props;
+
+    if (!_isEqual(previousItems, items)) {
```
Comment (alemikolo) [inline client/src/modules/list/components/Items/ItemsList.jsx]: Items were an array before. Is it necessary for them to be an object? Can we use array and compare items length? Isn't it simpler and faster? And without external libs?

File: client/src/modules/list/model/actionTypes.js [hunk 11]
```diff
@@ -22,8 +22,7 @@ export const ListActionTypes = enumerable('list')(
   ...asyncTypes('REMOVE_MEMBER'),
   ...asyncTypes('REMOVE_OWNER_ROLE'),
   ...asyncTypes('RESTORE'),
-  ...asyncTypes('UPDATE'),
-  ...asyncTypes('UPDATE_LIMIT')
+  ...asyncTypes('UPDATE')
```
Comment (alemikolo) [inline client/src/modules/list/model/actionTypes.js]: restore `...asyncTypes('UPDATE_LIMIT')`

File: client/src/modules/list/model/actions.js [hunk 12]
```diff
@@ -177,14 +177,6 @@ const fetchAvailableListsFailure = payload => ({
...
 
-const updateLimitSuccess = payload => ({
-  type: ListActionTypes.UPDATE_LIMIT_SUCCESS,
-  payload
-});
-
-export const updateLimit = (limit, listId) => dispatch =>
-  dispatch(updateLimitSuccess({ limit, listId }));
-
 export const fetchListData = listId => dispatch =>
```
Comment (alemikolo) [inline client/src/modules/list/model/actions.js]: restore
```
const updateLimitSuccess = payload => ({	
  type: ListActionTypes.UPDATE_LIMIT_SUCCESS,	
  payload	
});	

export const updateLimit = (limit, listId) => dispatch =>	
  dispatch(updateLimitSuccess({ limit, listId }));
```

File: client/src/modules/list/model/reducer.js [hunk 13]
```diff
@@ -317,21 +317,6 @@ const lists = (state = {}, action) => {
         '_id'
       );
     }
```
Comment (alemikolo) [inline client/src/modules/list/model/reducer.js]: restore
```
case ListActionTypes.UPDATE_LIMIT_SUCCESS: {	
      const {	
        payload: { listId, limit }	
      } = action;	

      const updatedState = { ...state[listId].listState, ...limit };	

      return {	
        ...state,	
        [listId]: {	
          ...state[listId],	
          listState: updatedState	
        }	
      };	
    }
```

File: client/src/modules/list/components/ItemsContainer.jsx (hunk 7 above)
Comment (adamklepacz) [reply client/src/modules/list/components/ItemsContainer.jsx]: I think it should stay plural cause it's about ITEMS not Item count. 
We count items, not a single item.

but I'm not a linguist :D

File: client/src/locales/en.json (hunk 5 above)
Comment (marekrozmus) [reply client/src/locales/en.json]: use https://github.com/formatjs/react-intl/blob/master/docs/Components.md#message-syntax

File: client/src/modules/list/components/ItemsContainer.jsx [hunk 14]
```diff
@@ -9,8 +9,11 @@ import ItemsList from 'modules/list/components/Items';
 import { getCurrentUser } from 'modules/user/model/selectors';
 import { IntlPropType } from 'common/constants/propTypes';
 import './ItemsContainer.scss';
-import Filter from 'common/components/Filter';
+import Badge from 'common/components/Badge';
 import { CloseIcon } from 'assets/images/icons';
+import Filter from 'common/components/Filter';
+
+const badgeContext = 'items';
```
Comment (marekrozmus) [inline client/src/modules/list/components/ItemsContainer.jsx]: how do you want to translate this?

Comment (marekrozmus) [reply client/src/modules/list/components/ItemsContainer.jsx]: use https://github.com/formatjs/react-intl/blob/master/docs/Components.md#message-syntax inside the json locale file

File: client/src/modules/list/components/Items/ItemsList.jsx [hunk 15]
```diff
@@ -26,19 +26,19 @@ class ItemsList extends PureComponent {
...
     const { items } = this.props;
 
     if (previousItems.length !== items.length) {
-      this.handleItems();
+      this.prepareDisplayedItems();
     }
   }
 
-  handleItems = () => {
+  prepareDisplayedItems = () => {
```
Comment (alemikolo) [inline client/src/modules/list/components/Items/ItemsList.jsx]: handleItemsToDisplay

Comment (adamklepacz) [reply client/src/modules/list/components/Items/ItemsList.jsx]: I think handle is more for handle an action from user input, button press etc.

File: client/src/modules/list/components/ItemsContainer.jsx (hunk 7 above)
Comment (Ciunkos) [reply client/src/modules/list/components/ItemsContainer.jsx]: https://english.stackexchange.com/a/11352 and https://english.stackexchange.com/questions/1314/user-accounts-or-users-account as a ref.

File: client/src/locales/en.json [hunk 16]
```diff
@@ -151,6 +151,7 @@
   "dashboard.index.add": "Add new sack",
   "list.archived-items-container.hide": "Hide archived items",
   "list.archived-items-container.show": "Show archived items",
+  "list.badge-context": "{totalCount, plural, =0 {''} one {item} other {items}}",
```
Comment (marekrozmus) [inline client/src/locales/en.json]: change name as it is `badge-context` but only for `items`

File: client/src/modules/list/components/ItemsContainer.jsx (hunk 7 above)
Comment (adamklepacz) [reply client/src/modules/list/components/ItemsContainer.jsx]: yes, you're right :)

File: client/src/common/components/Badge.jsx [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+import './Badge.scss';
+
+import { IntlPropType } from 'common/constants/propTypes';
+
+const Badge = ({
+  context,
+  count,
+  icon,
+  intl: { formatMessage },
+  totalCount
```
Comment (Ciunkos) [inline client/src/common/components/Badge.jsx]: Consider a rename to just `total`.

File: client/src/common/components/Badge.jsx [hunk 18]
```diff
@@ -0,0 +1,35 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { injectIntl } from 'react-intl';
+import './Badge.scss';
```
Comment (Ciunkos) [inline client/src/common/components/Badge.jsx]: Please move it to the internal import group.

File: client/src/common/components/Badge.scss [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+  border-radius: 16px;
+
+  &__icon {
+    margin-right: 4px;
+    display: flex;
+    align-items: center;
+    justify-content: center;
+
+    svg {
+      height: 14px;
```
Comment (Ciunkos) [inline client/src/common/components/Badge.scss]: Please use constant size helpers.

File: client/src/modules/list/components/Items/ItemsList.jsx [hunk 20]
```diff
@@ -12,41 +12,45 @@ import MessageBox from 'common/components/MessageBox';
...
-      disableItemAnimations(itemId, listId);
+    if (previousItems.length !== items.length) {
+      this.prepareDisplayedItems();
     }
+  }
+
+  prepareDisplayedItems = () => {
+    const { limit } = this.state;
+    const { items } = this.props;
+    const displayedItems = items.slice(0, limit).map(item => item);
```
Comment (Ciunkos) [inline client/src/modules/list/components/Items/ItemsList.jsx]: Consider rename `displayed` to `display` - if you apply this, please make sure to rename methods as well (`prepareDisplayItems` etc.).

File: client/src/modules/list/components/Items/ItemsList.jsx [hunk 21]
```diff
@@ -12,41 +12,45 @@ import MessageBox from 'common/components/MessageBox';
...
+
+    this.setState({ displayedItems }, this.displayedItemsCount);
```
Comment (Ciunkos) [inline client/src/modules/list/components/Items/ItemsList.jsx]: Methods must be named with verbs - `updateDisplayItemCount`.

File: client/src/modules/list/components/Items/ItemsList.jsx [hunk 22]
```diff
@@ -81,7 +115,7 @@ class ItemsList extends PureComponent {
     return archived ? (
       <ul className="items-list">
         <TransitionGroup component={null}>
-          {items.slice(0, limit).map(item => (
+          {displayedItems.map(item => (
```
Comment (Ciunkos) [inline client/src/modules/list/components/Items/ItemsList.jsx]: Please revert to purely based slicing if possible.

File: client/src/modules/list/components/ItemsContainer.jsx [hunk 23]
```diff
@@ -85,15 +90,23 @@ class ItemsContainer extends Component {
...
           </div>
         </header>
         {children}
         <div className="items__body">
           <ItemsList
             archived={archived}
-            done={done}
             isMember={isMember}
             items={items}
+            updateItemsCount={this.handleDisplayedItemsCount}
```
Comment (Ciunkos) [inline client/src/modules/list/components/ItemsContainer.jsx]: Please name handle<Event> and on<Event> naming pattern.

File: client/src/modules/list/components/Items/ItemsList.jsx (hunk 22 above)
Comment (adamklepacz) [reply client/src/modules/list/components/Items/ItemsList.jsx]: Not possible with the current implementation. 
I've created a task to reimplement it, to pass already sliced items from parent component.

**Comment by adamklepacz** [comment]
@przemyslawzalewski  check my solution now

File: client/src/modules/list/components/Items/ItemsList.jsx [hunk 24]
```diff
@@ -70,6 +62,43 @@ class ItemsList extends PureComponent {
     updateLimit(limit, listId);
   };
 
+  updateDisplayItemsCount = () => {
+    const { onUpdateItemsCount, items } = this.props;
+    const { limit } = this.state;
+    let displayedItemsCount;
```
Comment (Ciunkos) [inline client/src/modules/list/components/Items/ItemsList.jsx]: `Math.min(items.length, limit)`

**Pull request merged**